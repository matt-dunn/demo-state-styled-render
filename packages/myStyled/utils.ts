/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import Stylis from "stylis";

import { AnyRules, StylesheetPartial } from "./stylesheet";

const CLASS_NAME_PREFIX = "ms__";

const DEBUG = process.env.DEBUG === "true";

function isFunction(value: any): value is (...args: any[]) => any {
  return typeof value === "function";
}

export const parseRule = <T>(
  strings: TemplateStringsArray,
  args?: T[],
  props?: any
): string =>
  strings
    .reduce((rule: string[], part: string, index: number) => {
      rule.push(part);

      const arg = args?.[index];
      const value = arg && isFunction(arg) ? arg(props) : arg;

      if (value) {
        rule.push(value);
      }

      return rule;
    }, [])
    .join("");

// Use a fixed class prefix to simplify client/server class names
export const generateClassName = (hash: string): string =>
  `${CLASS_NAME_PREFIX}${hash}`;

const stylis = new Stylis({
  global: false,
});

export const updateSheetRule = (
  sheet: StylesheetPartial<CSSRuleList | AnyRules>,
  className: string,
  rule: string
) => {
  const DEBUG_RULES: string[] = [];

  const selectorText = `.${className}`;

  // See https://github.com/thysultan/stylis.js#plugins for plugin details
  stylis.use((context, content, selectors, parent) => {
    const selector = selectors[0];
    if (selector) {
      // Remove any additional specificity...
      // TODO: a bit hacky... can likely be improved/made more robust. this is intended to ensure child blocks are not inserted without the parent wrapper (e.g. @media rules)
      const normalizedSelector = `${parent[0] || ""}${
        (parent[0] &&
          selector.toString().replace(new RegExp(parent[0], "g"), "")) ||
        ""
      }`;

      // Do not include any global styles... Should be handled... globally! 2 = selector block, 3 = @at-rule block
      if (
        ((context === 2 && selector.startsWith(selectorText)) ||
          context === 3) &&
        normalizedSelector !== parent[0] &&
        selector.toLowerCase().indexOf(":global") === -1
      ) {
        // Add class postfix to localise animation name if required
        const currentSelector = `${selector}${
          (selector.toLowerCase().startsWith("@keyframes ") &&
            `-${className}`) ||
          ""
        }`;
        // Set the correct content if font-face
        const currentContent =
          (selector.toLowerCase().startsWith("@font-face") &&
            content.substr(1, content.length - 2)) ||
          content;

        sheet.insertRule(`${currentSelector} {${currentContent}}`);

        if (DEBUG) {
          DEBUG_RULES.push(`${currentSelector} {\n    ${content}\n  }`);
        }
      }
    }
  });

  stylis(selectorText, rule);

  if (DEBUG) {
    console.log(
      `UPDATE(${sheet.rules.length} rules)\n `,
      DEBUG_RULES.join("\n  ")
    );
  }

  return className;
};
