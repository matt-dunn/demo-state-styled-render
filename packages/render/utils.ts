/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import { Node } from "./types";

type MappedProps = {
  [key: string]: string;
};

const mappedProps: MappedProps = {
  className: "class",
  htmlFor: "for",
};

export const propertyMap = (name: string) => {
  if (mappedProps[name]) {
    return mappedProps[name];
  } else if (/^on/.test(name)) {
    return name.toLowerCase();
  }

  return name;
};

export const isNode = (node: any): node is Node =>
  (typeof node === "object" &&
    Object.prototype.hasOwnProperty.call(node, "type")) ||
  false;

export const classnames = (...classNames: (string | undefined)[]): string =>
  classNames.join(" ");

type LooseRef<T = any> = {
  current: T;
};

export const looseRef = <T>(initial: T): LooseRef<T> => ({
  current: initial,
});
