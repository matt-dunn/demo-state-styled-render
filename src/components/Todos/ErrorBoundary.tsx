/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

/** @jsx jsx **/

import { Node, jsx, useError } from "packages/render";

type ErrorBoundaryProps = {
  children: Node;
};

export const ErrorBoundary = ({ children }: ErrorBoundaryProps) => {
  const error = useError((error) => {
    console.log("HANDLE", error.message);
    // return { message: "wrapped: " + error.message };
  });

  if (error) {
    return <code>TODO ERROR BOUNDARY: {JSON.stringify(error)}</code>;
  }

  return children;
};
