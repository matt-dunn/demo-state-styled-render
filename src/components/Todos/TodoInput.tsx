/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

/** @jsx jsx **/

import { jsx, useState } from "packages/render";
import myStyled from "packages/myStyled";

import { Actions as TodoActions } from "./duck";

type TodoInputProps = {
  createTodo: TodoActions["createTodo"];
};

// Quick fix to allow stylelint to do it's thing:
const styled = myStyled;

const TodoListForm$ = styled("form")`
  display: flex;
  align-items: center;
  margin-bottom: 0.75rem;

  > label {
    display: flex;
    flex-grow: 1;
    align-items: center;
    margin: 0;
  }

  input[type="text"] {
    width: 100%;
    margin: 0 0.5rem 0 0.35rem;
  }
`;

export const TodoInput = ({ createTodo }: TodoInputProps) => {
  const [value, setValue] = useState("");

  return (
    <TodoListForm$
      onSubmit={(e: any) => {
        e.preventDefault();
        if (value.trim()) {
          createTodo(value.trim());
          setValue("");
        }
      }}
    >
      <label>
        Todo
        <input
          value={value}
          type="text"
          placeholder="Add new todo"
          className="form-control"
          onInput={(e: any) => {
            setValue(e.target.value);
          }}
        />
      </label>
      <button
        type="submit"
        className="btn btn-primary"
        disabled={!value.trim()}
      >
        Add
      </button>
    </TodoListForm$>
  );
};
