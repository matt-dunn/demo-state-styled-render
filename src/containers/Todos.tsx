/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

/** @jsx jsx **/

import { jsx } from "packages/render";
import { connect } from "packages/state";
import { Suspense } from "packages/render/Suspense";
import { Delay } from "packages/render/Delay";
import { lazy } from "packages/render/lazy";

import { AppState } from "../store";

import {
  Actions as TodoActions,
  actions as todoActions,
  TodoItems,
  TodoList,
} from "../components/Todos";
import { SubtleLoader } from "../components/Loader";

const TestSVG = lazy(
  () => import("../components/TestSVG"),
  (module) => module.TestSVG
);

type TodosProps = {
  todos: TodoItems;
  createTodo: TodoActions["createTodo"];
  deleteTodo: TodoActions["deleteTodo"];
  updateTodo: TodoActions["updateTodo"];
};

const Fallback = () => (
  <Delay delay={0}>
    <SubtleLoader>Loading…</SubtleLoader>
  </Delay>
);

const FallbackOuter = ({ pendingCount }: { pendingCount: number }) => (
  <Delay>
    <SubtleLoader>Loading outer…[{pendingCount}]</SubtleLoader>
  </Delay>
);

const Todos = ({ todos, createTodo, deleteTodo, updateTodo }: TodosProps) => (
  <main className="d-flex flex-md-row flex-column">
    <TodoList
      todos={todos}
      createTodo={createTodo}
      deleteTodo={deleteTodo}
      updateTodo={updateTodo}
      className="mr-md-2 mb-3 flex-grow-1 align-self-start col-md"
    />
    <div className="ml-md-2 flex-grow-1 col-md">
      <pre style={{ fontSize: "0.85rem" }}>
        {JSON.stringify(todos, undefined, 1)}
      </pre>
      {/*<Suspense Fallback={FallbackOuter}>*/}
      {/*  {todos.length >= 0 && (*/}
      {/*    <div*/}
      {/*      style={{*/}
      {/*        height: "50px",*/}
      {/*        marginBottom: "1rem",*/}
      {/*        display: "flex",*/}
      {/*      }}*/}
      {/*    >*/}
      {/*      <Suspense Fallback={Fallback}>*/}
      {/*        <TestSVG empty={todos.length === 0} />*/}
      {/*      </Suspense>*/}
      {/*      <TestSVG empty={todos.length === 0} />*/}
      {/*      <TestSVG empty={todos.length === 0} />*/}
      {/*    </div>*/}
      {/*  )}*/}
      {/*</Suspense>*/}
    </div>
  </main>
);

const ConnectedTodos = connect(
  ({ todos }: AppState) => ({
    todos,
  }),
  {
    ...todoActions,
  }
)(Todos);

export { ConnectedTodos as Todos };
