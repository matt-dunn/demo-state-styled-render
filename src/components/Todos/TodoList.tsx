/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

/** @jsx jsx **/

import { jsx } from "packages/render";
import myStyled from "packages/myStyled";

import { ErrorBoundary } from "./ErrorBoundary";

import { Todos } from "./Todos";
import { TodoItems } from "./types";
import { Actions as TodoActions } from "./duck";
import { Todo } from "./Todo";
import { TodoInput } from "./TodoInput";

type TodoListProps = {
  todos: TodoItems;
  createTodo: TodoActions["createTodo"];
  deleteTodo: TodoActions["deleteTodo"];
  updateTodo: TodoActions["updateTodo"];
  className?: string;
};

// Quick fix to allow stylelint to do it's thing:
const styled = myStyled;

const TodoList$ = styled("section")`
  padding: 1rem;
  width: 100%;
  border: 1px solid var(--secondary-color);
  border-radius: 0.25rem;
`;

const TodoListHeader$ = styled("header")`
  border-bottom: 1px solid var(--secondary-color);
  margin-bottom: 1rem;
`;

const Todos$ = styled(Todos)`
  ${({ todos }) =>
    (todos.length > 0 &&
      `
    border: 1px dotted var(--secondary-color);
    padding: 0.5rem;
  `) ||
    ""}
`;

export const TodoList = ({
  todos,
  createTodo,
  deleteTodo,
  updateTodo,
  className,
}: TodoListProps) => (
  <TodoList$ className={className}>
    {todos.length > 3 && (
      <TodoInput
        createTodo={createTodo}
        placeholder="Conditional component (>3 items)"
      />
    )}
    <TodoListHeader$>
      <ErrorBoundary>
        <TodoInput createTodo={createTodo} autoFocus={true} />
      </ErrorBoundary>
      <p>
        You have <strong>{todos.length}</strong> todo
        {(todos.length !== 1 && "s") || ""}
      </p>
    </TodoListHeader$>
    {todos.length > 4 && (
      <TodoInput
        createTodo={createTodo}
        placeholder="Conditional component (>4 items)"
      />
    )}
    <main>
      <Todos$ todos={todos}>
        {(todo) => (
          <Todo todo={todo} deleteTodo={deleteTodo} updateTodo={updateTodo} />
        )}
      </Todos$>
    </main>
  </TodoList$>
);
