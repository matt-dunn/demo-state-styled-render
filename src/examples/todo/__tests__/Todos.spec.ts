/** !
 * Copyright (c) 2020, Matt Dunn
 *
 * @author Matt Dunn (https://matt-dunn.github.io/)
 * @licence MIT
 */

import {mount} from "test/utils";

import {Todos} from "../Todos";

describe("Todos: Example of using mount", () => {
  it("should render component", async () => {
    const t: any = jest.fn();

    const wrapper = mount(Todos({todos: [{id: "1", text:"My Todo Item", complete: true}], children: t}));

    expect(wrapper.getDOMNode()).toMatchSnapshot();
  });
});
