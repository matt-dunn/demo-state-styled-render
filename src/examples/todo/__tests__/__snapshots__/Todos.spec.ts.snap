// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todos: Example of using mount should render component 1`] = `
<ul
  class=" ms__hUhVje"
>
  <li
    class=" ms__hsUUbt"
  />
</ul>
`;
